src/webui/api/searchcontroller.cpp

```diff
@@ -0,0 +1,407 @@
+    for (const QString &name : names)
+        SearchPluginManager::instance()->uninstallPlugin(name.trimmed());
+}
+
+void SearchController::enablePluginAction()
+{
+    checkParams({"names", "enable"});
+
+    const QStringList names = params()["names"].split('|');
+    const bool enable = Utils::String::parseBool(params()["enable"].trimmed(), false);
+
+    for (const QString &name : names)
+        SearchPluginManager::instance()->enablePlugin(name.trimmed(), enable);
+}
+
+void SearchController::updatePluginsAction()
+{
+    SearchPluginManager *pluginManager = SearchPluginManager::instance();
+
+    connect(pluginManager, &SearchPluginManager::checkForUpdatesFinished, this, &SearchController::checkForUpdatesFinished);
+    connect(pluginManager, &SearchPluginManager::checkForUpdatesFailed, this, &SearchController::checkForUpdatesFailed);
+    pluginManager->checkForUpdates();
+}
+
+void SearchController::searchTimeout(const QString &id)
+{
+    ISession *session = sessionManager()->session();
+    auto searchHandlers = session->getData(SEARCH_HANDLERS).value<SearchHandlerDict>();
+
+    if (searchHandlers.contains(id)) {
+        LogMsg(tr("Deleting search session %1 from web session due to timeout").arg(id), Log::INFO);
+        SearchHandlerPtr searchHandler = searchHandlers[id];
+        searchHandler->cancelSearch();
+        searchHandlers.remove(id);
+        session->setData(SEARCH_HANDLERS, QVariant::fromValue(searchHandlers));
+
+        removeActiveSearch(session, id);
+    }
+}
+
+void SearchController::checkForUpdatesFinished(const QHash<QString, PluginVersion> &updateInfo)
+{
+    if (updateInfo.isEmpty()) {
+        LogMsg(tr("All plugins are already up to date."), Log::INFO);
+        return;
+    }
+
+    LogMsg(tr("Updating %1 plugins").arg(updateInfo.size()), Log::INFO);
+
+    SearchPluginManager *pluginManager = SearchPluginManager::instance();
+    for (const QString &pluginName : updateInfo.keys()) {
+        LogMsg(tr("Updating plugin %1").arg(pluginName), Log::INFO);
+        pluginManager->updatePlugin(pluginName);
+    }
+}
+
+void SearchController::checkForUpdatesFailed(const QString &reason)
+{
+    LogMsg(tr("Failed to check for plugin updates: %1").arg(reason), Log::INFO);
+}
+
+void SearchController::searchFinished(ISession *session, const QString &id)
+{
+    removeActiveSearch(session, id);
+}
+
+void SearchController::searchFailed(ISession *session, const QString &id)
+{
+    removeActiveSearch(session, id);
+}
+
+QString SearchController::generateSearchId() const
+{
+    QString id;
+    auto searchHandlers = sessionManager()->session()->getData(SEARCH_HANDLERS).value<SearchHandlerDict>();
+
+    do {
+        const size_t size = 2;
+        quint32 tmp[size];
+
+        for (size_t i = 0; i < size; ++i)
+            tmp[i] = Utils::Random::rand();
+
+        id = QByteArray::fromRawData(reinterpret_cast<const char *>(tmp), sizeof(quint32) * size).toBase64();
```

Comment:
Why this needs to be so complicated? Why is the id string first of all? Could you explain, please, why a simple interger number does not suit here?

Reply:
> Why is the id string first of all?

I was responding to this question.

Reply:
Errrr... Thank you, of course, but why does the I'd need to be that complicated?

Reply:
I've switched the id over to using an integer.